File: models/Gpt-3.5-turbo/cpp/code/problems/tree_algorithms/Tree_Diameter.cpp

#include <iostream>
#include <vector>

using namespace std;

vector<vector<int>> adj;
vector<bool> visited;
int farthestNode, maxDistance;

void dfs(int node, int distance) {
    visited[node] = true;
    if (distance > maxDistance) {
        maxDistance = distance;
        farthestNode = node;
    }
    for (int neighbor : adj[node]) {
        if (!visited[neighbor]) {
            dfs(neighbor, distance + 1);
        }
    }
}

int main() {
    int n;
    cin >> n;

    adj.resize(n + 1);
    visited.resize(n + 1, false);

    for (int i = 1; i < n; i++) {
        int a, b;
        cin >> a >> b;
        adj[a].push_back(b);
        adj[b].push_back(a);
    }

    farthestNode = 0;
    maxDistance = 0;
    dfs(1, 0); // Perform DFS from any node (e.g., node 1)

    visited.assign(n + 1, false);
    maxDistance = 0;
    dfs(farthestNode, 0); // Perform DFS from the farthest node

    cout << maxDistance << endl;

    return 0;
}

File: models/Gpt-3.5-turbo/py/code/tree_algorithms/Tree_Diameter.py

n = int(input())
adj = {i: [] for i in range(1, n+1)}

for _ in range(n-1):
    a, b = map(int, input().split())
    adj[a].append(b)
    adj[b].append(a)

def dfs(node, distance, visited):
    visited[node] = True
    max_distance = distance
    farthest_node = node

    for neighbor in adj[node]:
        if not visited[neighbor]:
            new_distance = distance + 1
            temp_distance, temp_node = dfs(neighbor, new_distance, visited)
            if temp_distance > max_distance:
                max_distance = temp_distance
                farthest_node = temp_node

    return max_distance, farthest_node

# Find the farthest node from any arbitrary starting node
start_node = 1
visited = {i: False for i in range(1, n+1)}
_, farthest_node = dfs(start_node, 0, visited)

# Find the diameter of the tree
visited = {i: False for i in range(1, n+1)}
diameter, _ = dfs(farthest_node, 0, visited)

print(diameter)
